runners/flink/src/test/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperatorTest.java

```diff
@@ -910,21 +892,16 @@ public void testEnsureStateCleanupOnFinalWatermark() throws Exception {
         operator.keyedStateInternals.state(
             stateNamespace, StateTags.bag(stateId, ByteStringCoder.of()));
     state.add(ByteString.copyFrom("userstate".getBytes(Charsets.UTF_8)));
+    // No timers have been set for cleanup
+    assertThat(testHarness.numEventTimeTimers(), is(0));
+    // State has been created
     assertThat(testHarness.numKeyedStateEntries(), is(1));
 
     // Generate final watermark to trigger state cleanup
     testHarness.processWatermark(
         new Watermark(BoundedWindow.TIMESTAMP_MAX_VALUE.plus(1).getMillis()));
 
     assertThat(testHarness.numKeyedStateEntries(), is(0));
-
-    // Close should not repeat state cleanup
```

Comment:
Keep this to check that the cleanup isn't repeated even when the watermark is repeated.

Reply:
I believe the cleanup ought to be repeated if there is new state. If there is none, then no cleanup will be performed because we keep track of the state descriptors created and clear that set on clean-up.